### PR TITLE
FEATURE/ElvaRivle/DfWCIPVU-v2

### PR DESCRIPTION
**Pull request by ElvaRivle**
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * implemented login (on successful login JWT is returned)
 * added login service, controller and routes
--------------------------------------------------------------------
 OTHER
 * README typo
 * new JWT_SECRET env variable
 * closed mongoose connection on seed failure
 * globally enabled JSON body parsing
--------------------------------------------------------------------

### CODE REVIEW

File: controllers/login.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
+const loginService = require("../services/login")
+
+
+const loginUser = async (req, res) => {
+    const { email, password } = { ...req.body }
+
+    if (email === undefined) {
```
Comment (becarselman) [inline controllers/login.js]: you can use !email and that will cover this case

File: controllers/login.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            error: "Password not provided"
+        })
+    }
+
+    try {
+        const user = await loginService.getUserByEmail(email)
+
+        if (user === null) {
+            return res.status(404).json({
+                error: "Email not found"
```
Comment (becarselman) [inline controllers/login.js]: Can we found better message. I will suggest to create utils file with all error messages for example "Not found" and than you can reuse in other places.

**Review by becarselman — COMMENTED** [review]
Just remove gitKeep files if they no longer needed

File: controllers/login.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+const loginService = require("../services/login")
+
+
+const loginUser = async (req, res) => {
```
Comment (becarselman) [inline controllers/login.js:4]: we discuss this already on meeting but I want to set as reminder. 

The "models" folder contains the data models, which define the structure and behavior of the data entities in the application.
The "repositories" folder contains the code for interacting with the database, such as CRUD (Create, Read, Update, Delete) operations, and database queries.
The "services" folder contains the business logic of the application. This includes any application-specific rules or algorithms that are used to process data or implement functionality.
The "controllers" folder contains the logic for handling HTTP requests and responses. Controllers receive requests from the client, process them using the services, and return a response to the client.
Finally, the "routes" folder contains the API endpoints and associated HTTP methods. Each route is mapped to a specific controller method, and is responsible for handling a specific request from the client.

File: services/login.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    return await User.findOne({ email: email })
+}
+
+async function checkPassword(user, password) {
+    return await bcrypt.compare(password, user.password)
+}
+
+function generateJWTToken(user) {
+    return jwt.sign({ email: user.email }, process.env.JWT_SECRET, { expiresIn: "1d" })
```
Comment (becarselman) [inline services/login.js]: create new file where you can read all env vars and than import where you need

**Comment by ElvaRivle** [comment]
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * all errors added to errors.js file
 * all env variables added to env.js file
 * added user repository
--------------------------------------------------------------------
 OTHER
 * rewrote login service and controller
 * controller made minimalistic, all logic moved to service
--------------------------------------------------------------------

File: controllers/login.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+const errors = require("../utils/errors")
+
+
+const loginUser = async (req, res) => {
+    try {
+        const { email, password } = { ...req.body }
+
+        const { userId, token } = await loginService.loginUser(email, password)
+
+        return res.json({
```
Comment (becarselman) [inline controllers/login.js]: I think you are missing status code here.

File: utils/env.js
Comment (becarselman) [inline utils/env.js:1]: Find some different name for this. My suggestion is configuration

File: db/models/User.js [hunk 6]
```diff
@@ -1,22 +1,20 @@
...
-            type: String,
-            required: true
-        }
-    });
+const UserSchema = new mongoose.Schema({
+    email: {
+        type: String,
+        required: true,
+        unique: true,
+        validate: [validator.isEmail, "Email in invalid format"]
```
Comment (becarselman) [inline db/models/User.js]: You can reuse your new error messages in error file. Otherwise add new message : D

File: controllers/login.js (hunk 5 above)
Comment (ElvaRivle) [reply controllers/login.js]: It defaults to 200, but I will explicitly write it

File: db/models/User.js (hunk 6 above)
Comment (ElvaRivle) [reply db/models/User.js]: yikes, forgot that one 😄

**Comment by ElvaRivle** [comment]
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
MAIN
* renamed `utils` folder to `configuration`
* added EMAIL_INVALID_FORMAT error message
--------------------------------------------------------------------
OTHER
* explicitly returned 200 on successful login
* fixed misc warnings about unused requires
--------------------------------------------------------------------

**Comment by ElvaRivle** [comment]
FEATURE/ElvaRivle/DfWCIPVU-v2

--------------------------------------------------------------------
 MAIN
 * when `nodemon` restarts, release previous connection back to pool
--------------------------------------------------------------------
 OTHER
 * capture SIGTERM as well
 * extracted app termination login in separate function
 * handling CTRL + C behaviour on Windows via `npm run` is a no-go, [reason](https://stackoverflow.com/a/74751308)

**Review by becarselman — APPROVED** [review]

**Review by AntMarkoLovric — APPROVED** [review]

**Pull request merged**